src/front/wgsl/number.rs

```diff
@@ -0,0 +1,438 @@
+use std::borrow::Cow;
+
+use super::{NumberError, Token};
+
+/// When using this type assume no Abstract Int/Float for now
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Number {
+    /// Abstract Int (-2^63 ≤ i < 2^63)
+    AbstractInt(i64),
+    /// Abstract Float (IEEE-754 binary64)
+    AbstractFloat(f64),
+    /// Concrete i32
+    I32(i32),
+    /// Concrete u32
+    U32(u32),
+    /// Concrete f32
+    F32(f32),
+}
+
+// TODO: when implementing Creation-Time Expressions, remove the ability to match the minus sign
+
+pub(super) fn consume_number(input: &str) -> (Token<'_>, &str) {
+    let res = if let Some((num, rest)) = parse(input) {
+        (num, rest)
+    } else {
+        (Err(NumberError::Invalid), input)
+    };
+
+    let num = match res.0 {
+        Ok(Number::AbstractInt(num)) => {
+            use std::convert::TryFrom;
+            i32::try_from(num)
+                .map(Number::I32)
+                .map_err(|_| NumberError::NotRepresentable)
+        }
+        Ok(Number::AbstractFloat(num)) => {
+            let num = num as f32;
+            if num.is_finite() {
+                Ok(Number::F32(num))
+            } else {
+                Err(NumberError::NotRepresentable)
+            }
+        }
+        num => num,
+    };
+    (Token::Number(num), res.1)
+}
+
+enum Kind {
+    Int(IntKind),
+    Float(FloatKind),
+}
+
+enum IntKind {
+    I32,
+    U32,
+}
+
+enum FloatKind {
+    F32,
+    F16,
+}
+
+// The following regexes (from the WGSL spec) will be matched:
+
+// int_literal:
+// | / 0                                                                [iu]?   /
+// | / [1-9][0-9]*                                                      [iu]?   /
+// | / 0[xX][0-9a-fA-F]+                                                [iu]?   /
+
+// decimal_float_literal:
+// | / 0                                                                [fh]    /
+// | / [1-9][0-9]*                                                      [fh]    /
+// | / [0-9]*               \.[0-9]+            ([eE][+-]?[0-9]+)?      [fh]?   /
+// | / [0-9]+               \.[0-9]*            ([eE][+-]?[0-9]+)?      [fh]?   /
+// | / [0-9]+                                    [eE][+-]?[0-9]+        [fh]?   /
+
+// hex_float_literal:
+// | / 0[xX][0-9a-fA-F]*    \.[0-9a-fA-F]+      ([pP][+-]?[0-9]+        [fh]?)? /
+// | / 0[xX][0-9a-fA-F]+    \.[0-9a-fA-F]*      ([pP][+-]?[0-9]+        [fh]?)? /
+// | / 0[xX][0-9a-fA-F]+                         [pP][+-]?[0-9]+        [fh]?   /
+
+// You could visualize the regex below via https://debuggex.com to get a rough idea what `extract_info` is doing
+// -?(?:0[xX](?:([0-9a-fA-F]+\.[0-9a-fA-F]*|[0-9a-fA-F]*\.[0-9a-fA-F]+)(?:([pP][+-]?[0-9]+)([fh]?))?|([0-9a-fA-F]+)([pP][+-]?[0-9]+)([fh]?)|([0-9a-fA-F]+)([iu]?))|((?:[0-9]+[eE][+-]?[0-9]+|(?:[0-9]+\.[0-9]*|[0-9]*\.[0-9]+)(?:[eE][+-]?[0-9]+)?))([fh]?)|((?:[0-9]|[1-9][0-9]+))([iufh]?))
+
+fn parse(input: &str) -> Option<(Result<Number, NumberError>, &str)> {
```

Comment:
It seems like this could just return `Result<Number, NumberError>`, and replace the `None` returns with `Err(NumberError::Invalid)`, and then `consume_number` could just use the `?` operator. Does that work?

Reply:
Since the rest is also returned I wasn't sure if it was the responsibility of `parse` to give back the full input string. Another option would be: `(Result<Number, NumberError>, Option<&str>)`. What do you think?

Reply:
I tried making this change myself and realized I'd missed a bunch of details. There are really two levels of error here:

1) We may fail to recognize a numeric literal at all. In this case, the token returned by `consume_number` is `Token::Number(Err(NumberError::Invalid))`, and we don't advance the input.

2) We may recognize a numeric literal, but it may not be representable, or it may be an unimplemented type like `f16`. In this case, the token returned by `consume_number` is `Token::Number(Err(NumberError:: something else))`, and we do advance the input to the next token.

In the long term, we would like Naga to be able to return more than one error message at a time. I think that means that, since we're here, `parse` should always return a rest string, even when we fail to recognize any literal at all, and try to advance us across the initial run of characters that looked okay. (There should always be such an initial run, or else the lexer wouldn't have called `consume_number`.)

So I think `parse` should return `(Result<Number, NumberError>, &str)`, and then `consume_number` should look like this:

```
pub(super) fn consume_number(input: &str) -> (Token<'_>, &str) {
    let (result, rest) = parse(input);
    (Token::Number(result.and_then(Number::abstract_to_concrete)), rest)
}
```

where:

```
impl Number {
    fn abstract_to_concrete(self) -> Result<Number, NumberError> {
        match self {
            Number::AbstractInt(num) => {
                use std::convert::TryFrom;
                i32::try_from(num)
                    .map(Number::I32)
                    .map_err(|_| NumberError::NotRepresentable)
            }
            Number::AbstractFloat(num) => {
                let num = num as f32;
                if num.is_finite() {
                    Ok(Number::F32(num))
                } else {
                    Err(NumberError::NotRepresentable)
                }
            }
            num => Ok(num),
        }
    }
}
```

And then instead of `return None`, `parse` would say something like:

```
return (Err(NumberError::Invalid), rest_to_str!(bytes));
```